src/app/ReadClient.cpp

```diff
@@ -957,48 +965,99 @@ CHIP_ERROR ReadClient::SendSubscribeRequestImpl(const ReadPrepareParams & aReadP
                                                     Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse)));
 
-    mPeerNodeId  = aReadPrepareParams.mSessionHolder->AsSecureSession()->GetPeerNodeId();
-    mFabricIndex = aReadPrepareParams.mSessionHolder->GetFabricIndex();
-
+    mPeer = aReadPrepareParams.mSessionHolder->AsSecureSession()->GetPeer();
     MoveToState(ClientState::AwaitingInitialReport);
 
     return CHIP_NO_ERROR;
 }
 
-void ReadClient::OnResubscribeTimerCallback(System::Layer * apSystemLayer, void * apAppState)
+CHIP_ERROR ReadClient::DefaultResubscribePolicy(CHIP_ERROR aTerminationCause)
 {
-    ReadClient * const _this = reinterpret_cast<ReadClient *>(apAppState);
-    assert(_this != nullptr);
-    _this->SendSubscribeRequest(_this->mReadPrepareParams);
-    _this->mNumRetries++;
+    VerifyOrReturnError(IsIdle(), CHIP_ERROR_INCORRECT_STATE);
+
+    auto timeTillNextResubscription = ComputeTimeTillNextSubscription();
+    ChipLogProgress(DataManagement,
+                    "Will try to resubscribe to %02x:" ChipLogFormatX64 " at retry index %" PRIu32 " after %" PRIu32
+                    "ms due to error %" CHIP_ERROR_FORMAT,
+                    GetFabricIndex(), ChipLogValueX64(GetPeerNodeId()), mNumRetries, timeTillNextResubscription,
+                    aTerminationCause.Format());
+    ReturnErrorOnFailure(ScheduleResubscription(timeTillNextResubscription, aTerminationCause == CHIP_ERROR_TIMEOUT));
+    return CHIP_NO_ERROR;
 }
 
-bool ReadClient::ResubscribeIfNeeded(uint32_t & aNextResubscribeIntervalMsec)
+void ReadClient::HandleDeviceConnected(void * context, OperationalDeviceProxy * device)
 {
-    bool shouldResubscribe       = true;
-    uint32_t intervalMsec        = 0;
-    aNextResubscribeIntervalMsec = 0;
-    if (mReadPrepareParams.mResubscribePolicy == nullptr)
+    ReadClient * const _this = static_cast<ReadClient *>(context);
+    VerifyOrDie(_this != nullptr);
+
+    _this->mReadPrepareParams.mSessionHolder.Release();
+    _this->mReadPrepareParams.mSessionHolder.Grab(device->GetSecureSession().Value());
+
+    auto err = _this->SendSubscribeRequest(_this->mReadPrepareParams);
+    if (err != CHIP_NO_ERROR)
     {
-        ChipLogDetail(DataManagement, "mResubscribePolicy is null");
-        return false;
+        _this->Close(err);
     }
-    mReadPrepareParams.mResubscribePolicy(mNumRetries, intervalMsec, shouldResubscribe);
-    if (!shouldResubscribe)
+}
+
+void ReadClient::HandleDeviceConnectionFailure(void * context, PeerId peerId, CHIP_ERROR err)
+{
+    ReadClient * const _this = static_cast<ReadClient *>(context);
+    VerifyOrDie(_this != nullptr);
+
+    ChipLogError(DataManagement, "Failed connecting to device with error '%s'", err.AsString());
+
+    auto * caseSessionManager = InteractionModelEngine::GetInstance()->GetCASESessionManager();
+    VerifyOrDie(caseSessionManager != nullptr);
```

Comment:
Discussed offline.  I would request that we instead propagate an error here if we are claiming that it can be legal to pass a null CASESessionManager.

Reply:
I'll convert the other VerifyOrDie to an error, but this one should likely still be a VerifyOrDie, since to get to this callback, it meant that you had to have had a CASESessionManager housing OperationalDeviceProxy instances.

Reply:
I would like to request then that this feature be moved out of the ReadClient base class and into an inherited class.

We already have consensus that ReadClient is one place where re-initiating CASE can occur, but that it's equally reasonable for that to happen in the app.  So let's make the code promise to support that and formalize that promise with this feature in a class extension.  Right now, there is no stability to this promise going forward and it would be non-obvious to future maintainers that ReadClient must support a usage without these calls into CASESessionManager.

I believe this change can be made without significant redesign of this feature.

Reply:
The existing ReadClient already handles re-subscriptions, so teasing that out into a separate extended class would be taking away a feature that already exists, which I don't think is something we can do.

Given that, it's not trivial nor sensible to split apart the CASE re-establishment logic from the timer management and re-sub scheduling logic that already exists in the client today into a separate extended class. If CASE is requested, the callback fired by the timer has to establish CASE, and not just motor on with the handle present inside the client object, which is why both bits of logic have to be sandwiched together.

> it would be non-obvious to future maintainers that ReadClient must support a usage without these calls into CASESessionManager.

Wouldn't documentation help here? Isn't reading that documentation an expectation of any future maintainer?

I'd view all of this from the lens of actually _fixing_ our current re-sub logic - it's broken because it never actually successfully re-subscribes in the 99% of cases that matter, because it never attempts to re-do CASE. Against that lens, this PR is not a new feature - it's a bug fix :)

Reply:
That is a fair assessment.  You're right that resubscription as a feature in the ReadClient isn't new.  It just didn't work.

Based on this then, I would welcome any documentation and test cases that give stability around supporting usage without a CASESessionManager.  It's just very important to us that this be a fully supported usage.

Reply:
Added a test that validates re-subscription without a CASESessionManager (this is in fact that the only thing we can test easily right now without adding a lot of extra test infra).